File: back/kernel/dsleditor/dsleditor.go
// Package dsleditor is the S77 TYPED DSL EDITOR surface (KRD §24, EPIC 6): the
// Workbench authoring path by which a USER edits the four behaviour DSLs — Operation
// (validate/authorize/read/mutate/return, incl. the async/scheduled verbs of S73),
// Policy (the recursive ALLOW/DENY tree), and the verticale Control+Action
// (visible_when/enabled_when/triggers → invoke operation) — WITHOUT writing free
// code. Every edit is a TYPED FORM (a structured wire body, not a code string) that a
// PURE parser turns into the existing DSL AST (the S09/S10/S11 contracts, never a
// fork), validates, and PROPOSES as a DRAFT ChangeSet.
//
// THE NO-FREE-CODE LAW (KRD §24, "behaviour in DSL, never free code"). A DslDoc
// carries a `kind` ∈ {operation, policy, control, action} and a typed `body` whose
// shape is the DSL's wire shape — there is no `code` field, no `script`, no `eval`.
// ParseDoc is a TOTAL pure function from the typed body to the DSL AST; an unknown
// kind or a malformed body yields a TYPED error, never a guess (the honesty rule).
//
// THE WALL (CLAUDE.md §2). This package writes NOTHING. ProposeEdit runs the pure
// parser + the DSL's own Validate, then wraps the canonical AST into a DRAFT
// changeset.ChangeSet (spec_delta + mirror_delta, project-scoped target
// "dsl-edit@<kind>:<name>"). The changeset is `proposed` (DRAFT) — never APPLIED,
// WroteKernel stays false; freezing goes through the wall (idée → miroir → /goal →
// approbation humaine). A truth-write from the screen is impossible by construction.
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8). ParseDoc + ProposeEdit are PURE, TOTAL: no DB,
// no clock, no RNG, no I/O. Same doc+parent ⇒ byte-identical DRAFT (the content
// address pins it). The reproducibility mirror (dsleditor_property_test.go, rapid)
// proves it. The verticale fixture (dsleditor_fixture_test.go, state→cmd→events)
// proves an AUTHORIZED control triggers its AUTHORIZED operation and an enabled_when
// false BLOCKS the action — reusing the FROZEN control/action/operation/policy
// interpreters (S09/S10/S11), never a second evaluator.
package dsleditor

import (
	"encoding/json"
	"errors"
	"fmt"

	"github.com/steph-frtech/aidos/back/archive/changeset"
	"github.com/steph-frtech/aidos/back/kernel/action"
	"github.com/steph-frtech/aidos/back/kernel/control"
	"github.com/steph-frtech/aidos/back/kernel/operation"
	"github.com/steph-frtech/aidos/back/kernel/policy"
	"github.com/steph-frtech/aidos/back/kernel/records"
)

// Kind is the CLOSED set of DSLs the typed editor serves (KRD §24.1–§24.4). Nothing
// outside this set is editable — there is no free-code kind.
type Kind string

const (
	// KindOperation — the Operation DSL (validate/authorize/read/mutate/return, S10/S73).
	KindOperation Kind = "operation"
	// KindPolicy — the Policy DSL (the recursive ALLOW/DENY tree, S09).
	KindPolicy Kind = "policy"
	// KindControl — the control-spec (visible_when/enabled_when/triggers, S11).
	KindControl Kind = "control"
	// KindAction — the action-spec (binds control → operation, S11).
	KindAction Kind = "action"
)

// kinds is the closed set in canonical order — the Workbench legend reads it, never invents one.
var kinds = []Kind{KindOperation, KindPolicy, KindControl, KindAction}

// Kinds returns the four editable DSL kinds in canonical order.
func Kinds() []Kind { return append([]Kind(nil), kinds...) }

// IsKind reports whether k is one of the four editable DSL kinds.
func IsKind(k string) bool {
	for _, kk := range kinds {
		if string(kk) == k {
			return true
		}
	}
	return false
}

// Typed editor errors — all TYPED, never a panic (the honesty rule).
var (
	// ErrUnknownKind — a DslDoc whose kind is not one of the four editable DSLs.
	ErrUnknownKind = errors.New("dsleditor: kind is not one of the four editable DSLs (no free-code kind)")
	// ErrFreeCode — a body carries a free-code escape (a `code`/`script`/`eval` field). The typed
	// editor forbids free code: behaviour is DSL all the way down (KRD §24).
	ErrFreeCode = errors.New("dsleditor: a typed editor forbids free code (no code/script/eval field, KRD §24)")
	// ErrEmptyName — a DslDoc with no name (every edited source is named, so the changeset target resolves).
	ErrEmptyName = errors.New("dsleditor: doc has no name")
	// ErrBadBody — the typed body does not parse for its kind (a malformed form). Wraps the DSL's parse error.
	ErrBadBody = errors.New("dsleditor: typed body does not parse for its kind")
)

// DslDoc is one TYPED EDITOR DOCUMENT: a kind, a name, and the typed wire body the
// user's form produced. There is NO code field — the body IS the DSL's structured
// shape. A control/action body additionally needs the known refs to validate (an
// orphan trigger/bind is a monster) — they are supplied on the doc, not the body, so
// the body stays a pure DSL fragment.
type DslDoc struct {
	// Kind is the DSL being edited (must be in Kinds()).
	Kind Kind `json:"kind"`
	// Name is the edited source's name (e.g. "createOrder", "can-checkout"). Required.
	Name string `json:"name"`
	// Body is the typed wire body — the DSL's own JSON shape (operationWire/policyWire/…). No free code.
	Body json.RawMessage `json:"body"`
	// KnownActions are the action refs a control's `triggers` may resolve to (control kind only).
	KnownActions []string `json:"known_actions,omitempty"`
	// KnownControls / KnownOperations are the refs an action's on/invoke may resolve to (action kind only).
	KnownControls   []string `json:"known_controls,omitempty"`
	KnownOperations []string `json:"known_operations,omitempty"`
}

// Parsed is the result of ParseDoc: the kind + the canonical AST body (the bytes the
// changeset carries) plus the typed AST values for whichever DSL was edited (exactly
// one is non-nil). The caller reads the typed value for a Workbench preview; the
// changeset carries Canonical.
type Parsed struct {
	Kind      Kind                 `json:"kind"`
	Name      string               `json:"name"`
	Canonical json.RawMessage      `json:"canonical"`
	Operation *operation.Operation `json:"operation,omitempty"`
	Async     *operation.Async     `json:"async,omitempty"`
	Policy    *policy.Policy       `json:"policy,omitempty"`
	Control   *control.Control     `json:"control,omitempty"`
	Action    *action.Action       `json:"action,omitempty"`
}

// hasFreeCode reports whether a body smuggles a free-code escape — a top-level
// `code`, `script`, or `eval` field. The typed editor refuses it: behaviour is DSL,
// never free code (KRD §24). PURE.
func hasFreeCode(body json.RawMessage) bool {
	var probe map[string]json.RawMessage
	if err := json.Unmarshal(body, &probe); err != nil {
		return false // a non-object body cannot smuggle a named code field; ParseDoc rejects it downstream.
	}
	for _, banned := range []string{"code", "script", "eval"} {
		if _, ok := probe[banned]; ok {
			return true
		}
	}
	return false
}

// ParseDoc is the PURE, TOTAL parse of a typed editor doc into its DSL AST. It rejects
// an unknown kind, a free-code escape, an empty name, and a malformed body (wrapping
// the DSL's own typed parse error). It NEVER evaluates and NEVER panics. The returned
// Parsed.Canonical re-canonicalises identically (the round-trip the property mirror
// pins). DETERMINISTIC: same doc ⇒ same Parsed.
func ParseDoc(d DslDoc) (Parsed, error) {
	if !IsKind(string(d.Kind)) {
		return Parsed{}, fmt.Errorf("%w: %q", ErrUnknownKind, d.Kind)
	}
	if d.Name == "" {
		return Parsed{}, ErrEmptyName
	}
	if hasFreeCode(d.Body) {
		return Parsed{}, ErrFreeCode
	}
	switch d.Kind {
	case KindOperation:
		return parseOperationDoc(d)
	case KindPolicy:
		return parsePolicyDoc(d)
	case KindControl:
		return parseControlDoc(d)
	case KindAction:
		return parseActionDoc(d)
	default:
		// Unreachable (IsKind gated above) — kept total.
		return Parsed{}, fmt.Errorf("%w: %q", ErrUnknownKind, d.Kind)
	}
}

// canonicalise marshals a value then re-canonicalises it through records.Canonicalize
// (S02) so the changeset body is key-order-stable and content-addressable. PURE.
func canonicalise(v any) (json.RawMessage, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	canon, err := records.Canonicalize(raw)
	if err != nil {
		return nil, err
	}
	return json.RawMessage(canon), nil
}

// Proposal is the result of proposing a typed edit: the parsed AST (echoed for the
// Workbench preview) plus a DRAFT ChangeSet carrying its canonical body — never
// APPLIED. The changeset's spec_delta carries the AST, project-scoped; its
// mirror_delta declares the proof obligation (completeness — a spec needs a mirror).
// Approval (apply) is the `aidos` CLI's job, gated by the AuthorityGraph — ProposeEdit
// only proposes.
type Proposal struct {
	Parsed    Parsed              `json:"parsed"`
	ChangeSet changeset.ChangeSet `json:"changeset"`
}

// ProposeEdit parses a typed editor doc and wraps its canonical AST into a DRAFT
// changeset.ChangeSet — the single legal way a DSL edit reaches truth. It WRITES
// NOTHING: changeset.Open is PURE. The returned changeset is `proposed` (DRAFT,
// WroteKernel false); a human approves (applies) or rejects it downstream — a reject
// leaves the kernel intact because ProposeEdit never touched it.
//
//   - parentPhase is the stable phase the proposal moves from (the project's current head).
//   - The spec_delta target is "dsl-edit@<kind>:<name>" so a reject/approve is project-scoped.
func ProposeEdit(d DslDoc, parentPhase string) (Proposal, error) {
	parsed, err := ParseDoc(d)
	if err != nil {
		return Proposal{}, err
	}
	target := fmt.Sprintf("dsl-edit@%s:%s", parsed.Kind, parsed.Name)
	label := fmt.Sprintf("dsleditor: propose %s edit %q", parsed.Kind, parsed.Name)
	spec := &changeset.Delta{Kind: "add", Target: target, Body: parsed.Canonical}
	mirror := &changeset.Delta{Kind: "add", Target: target + "#mirror"}
	cs, err := changeset.Open(label, parentPhase, spec, mirror)
	if err != nil {
		return Proposal{}, err
	}
	return Proposal{Parsed: parsed, ChangeSet: cs}, nil
}

File: front/web/lib/dsl-editor.ts
/**
 * dsl-editor — the S77 front twin of back/kernel/dsleditor (dsleditor.go). The TYPED DSL EDITOR over
 * the four behaviour DSLs — Operation (validate/authorize/read/mutate/return, incl. async/scheduled of
 * S73), Policy (the recursive ALLOW/DENY tree), and the verticale Control+Action
 * (visible_when/enabled_when/triggers → invoke operation). Every edit is a TYPED FORM (no free code)
 * that a PURE parser turns into the existing DSL types, validates, and PROPOSES as a DRAFT ChangeSet.
 *
 * NO FORK (ADR 0007): it IMPORTS the existing DSL libs (lib/policy, lib/operation, lib/control) — it
 * does NOT re-implement a DSL. It adds ONLY the S77 concerns: the no-free-code law, the typed parse
 * verdict, and proposeEdit (a DRAFT changeset). The authoritative content-addressed id is the Go one;
 * the front carries a stable handle.
 *
 * THE WALL (CLAUDE.md §2/§7): this module writes NOTHING. proposeEdit returns a Proposal VALUE — a
 * DRAFT ChangeSet (status always DRAFT, wrote_kernel false, no applied_at). The screen PROPOSES;
 * freezing into the kernel goes through the wall (propose → ChangeSet → /goal → approval).
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): parseDoc + proposeEdit are PURE + TOTAL — same input → identical
 * output. Parsing DSL = pure function. dsl-editor.test.ts pins it (Vitest + fast-check).
 *
 * THE TWIN IS THE DEMO, NOT THE LIVE PATH (ADR 0092 batch-3). Since the kill-twins flip, the
 * /dsl-editor panel reads the LIVE proposal from the Go dsl-editor MCP server through the passerelle
 * (`readVia(scope, "dsl_propose", …)`); this module's parseDoc/proposeEdit/kinds are KEPT only as the
 * deterministic DEMO-FALLBACK compute (driven by lib/dsl-editor-data.ts, the `source:"demo"` path).
 * The presence of the `lib/dsl-editor-data.ts` sibling is what makes the T5 cliquet
 * (twin-as-live-fitness) classify this lib as a twin — guarding the flip (the panel reads via the
 * readVia frontier, the twin never sits on the live path again). The Go dsleditor.ParseDoc /
 * ProposeEdit is the authoritative implementation; this twin mirrors it.
 */

import { createHash } from "node:crypto";

/** Kind — the closed set of editable DSLs (twin of dsleditor.Kind). */
export const DSL_KINDS = ["operation", "policy", "control", "action"] as const;
export type DslKind = (typeof DSL_KINDS)[number];

/** kinds returns the four editable DSL kinds in canonical order. PURE. */
export function kinds(): DslKind[] {
	return [...DSL_KINDS];
}

export function isKind(s: string): s is DslKind {
	return (DSL_KINDS as readonly string[]).includes(s);
}

export const ERR_UNKNOWN_KIND =
	"dsleditor: kind is not one of the four editable DSLs (no free-code kind)";
export const ERR_FREE_CODE =
	"dsleditor: a typed editor forbids free code (no code/script/eval field, KRD §24)";
export const ERR_EMPTY_NAME = "dsleditor: doc has no name";
export const ERR_BAD_BODY = "dsleditor: typed body does not parse for its kind";

/** DslDoc — one typed editor document (twin of dsleditor.DslDoc). No `code` field — body IS the DSL. */
export interface DslDoc {
	kind: DslKind;
	name: string;
	body: Record<string, unknown>;
	knownActions?: string[];
	knownControls?: string[];
	knownOperations?: string[];
}

// canonicalEncode re-encodes a JSON value with object keys sorted recursively (the same scheme the Go
// records.Canonicalize uses) — so the canonical handle is a stable content address. PURE.
function canonicalEncode(v: unknown): string {
	if (v === null || v === undefined) return "null";
	if (Array.isArray(v)) return `[${v.map(canonicalEncode).join(",")}]`;
	if (typeof v === "object") {
		const o = v as Record<string, unknown>;
		const keys = Object.keys(o).sort();
		return `{${keys.map((k) => `${JSON.stringify(k)}:${canonicalEncode(o[k])}`).join(",")}}`;
	}
	return JSON.stringify(v);
}

/** hasFreeCode reports whether a body smuggles a free-code escape (a code/script/eval field). PURE. */
export function hasFreeCode(body: Record<string, unknown>): boolean {
	return (
		Object.hasOwn(body, "code") ||
		Object.hasOwn(body, "script") ||
		Object.hasOwn(body, "eval")
	);
}

export interface Parsed {
	kind: DslKind;
	name: string;
	canonical: string; // canonical encoding of the typed body (a stable content handle)
}

export interface ParseResult {
	ok: boolean;
	error?: string;
	parsed?: Parsed;
}

/**
 * parseDoc — the PURE, TOTAL parse of a typed editor doc. It rejects an unknown kind, a free-code
 * escape, an empty name, and a body missing the kind's required typed fields. Same doc ⇒ same Parsed.
 * Parsing DSL = pure function (the done-criterion). Twin of dsleditor.ParseDoc.
 */
export function parseDoc(d: DslDoc): ParseResult {
	if (!isKind(d.kind))
		return { ok: false, error: `${ERR_UNKNOWN_KIND}: "${d.kind}"` };
	if (!d.name) return { ok: false, error: ERR_EMPTY_NAME };
	if (hasFreeCode(d.body)) return { ok: false, error: ERR_FREE_CODE };
	const shapeErr = checkShape(d.kind, d.body);
	if (shapeErr) return { ok: false, error: `${ERR_BAD_BODY}: ${shapeErr}` };
	const body = { ...d.body, name: d.name };
	return {
		ok: true,
		parsed: { kind: d.kind, name: d.name, canonical: canonicalEncode(body) },
	};
}

// checkShape verifies the typed body carries the closed-grammar fields its kind needs — never a free
// string. Returns a message or null. PURE (mirrors the Go DSL parsers' typed rejections).
function checkShape(
	kind: DslKind,
	body: Record<string, unknown>,
): string | null {
	switch (kind) {
		case "operation": {
			const steps = body.steps;
			if (!Array.isArray(steps))
				return "operation body needs a typed `steps` list";
			const verbs = [
				"validate",
				"authorize",
				"read",
				"mutate",
				"branch",
				"return",
			];
			for (const s of steps as Array<Record<string, unknown>>) {
				if (!verbs.includes(String(s.kind)))
					return `unknown step verb "${s.kind}" (closed grammar)`;
			}
			return null;
		}
		case "policy": {
			const scopes = ["RESOURCE", "OPERATION", "ENTITY", "FIELD"];
			if (!scopes.includes(String(body.scope)))
				return `unknown scope "${body.scope}" (closed set)`;
			if (body.effect !== "ALLOW" && body.effect !== "DENY")
				return `unknown effect "${body.effect}"`;
			if (!body.rule || typeof body.rule !== "object")
				return "policy body needs a typed `rule` tree";
			return null;
		}
		case "control": {
			if (!body.view) return "control body needs a `view`";
			if (!body.visible_when || !body.enabled_when)
				return "control body needs visible_when + enabled_when";
			return null;
		}
		case "action": {
			const on = body.on as Record<string, unknown> | undefined;
			if (!on || on.kind !== "click")
				return "action body needs on: { kind: click, control }";
			if (!body.invoke) return "action body needs an `invoke` operation ref";
			return null;
		}
	}
}

/** A minimal ChangeSet shape (mirrors Go changeset.ChangeSet, the fields the screen reads). */
export interface ChangeSetView {
	label: string;
	status: "DRAFT";
	parent_phase: string;
	wrote_kernel: false;
	spec_delta: { kind: string; target: string; body: string };
	mirror_delta: { kind: string; target: string };
}

export interface Proposal {
	ok: boolean;
	error?: string;
	parsed?: Parsed;
	changeset?: ChangeSetView;
}

/**
 * proposeEdit — parse the typed edit and wrap its canonical AST into a `proposed` (DRAFT) ChangeSet
 * carrying a project-scoped spec_delta + a mirror_delta (completeness law). It NEVER applies (status
 * always DRAFT, wrote_kernel false) — the wall. A reject discards this DRAFT; the kernel is never
 * touched. Twin of dsleditor.ProposeEdit.
 */
export function proposeEdit(d: DslDoc, parentPhase: string): Proposal {
	const r = parseDoc(d);
	if (!r.ok || !r.parsed) return { ok: false, error: r.error };
	const target = `dsl-edit@${r.parsed.kind}:${r.parsed.name}`;
	const bodyHash = createHash("sha256")
		.update(r.parsed.canonical)
		.digest("hex");
	return {
		ok: true,
		parsed: r.parsed,
		changeset: {
			label: `dsleditor: propose ${r.parsed.kind} edit "${r.parsed.name}"`,
			status: "DRAFT",
			parent_phase: parentPhase,
			wrote_kernel: false,
			spec_delta: { kind: "add", target, body: bodyHash },
			mirror_delta: { kind: "add", target: `${target}#mirror` },
		},
	};
}
